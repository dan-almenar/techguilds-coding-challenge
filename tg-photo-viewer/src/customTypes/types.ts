type Target = '' | '_blank';
export interface NavbarProps {
	data: { id: string, url: string, label: string }[];
};
export interface FooterProps {
	data: { year: number, url: string, label: string, target?: Target };
};
export interface ImageData {
	id: string,
	width: number,
	height: number,
	blur_hash: string,
	description: string,
	alt_description: string,
	location: { [key: string]: string },
	urls: {
		full: string,
		regular: string,
		small: string
	},
	links: {
		html: string,
		download: string
	},
	likes: number,
	user: {
		id: string,
		username: string,
		name: string,
		portfolio_url: string
	}
};
export interface RequestOptions {
	headers?: { [key: string]: string };
};
