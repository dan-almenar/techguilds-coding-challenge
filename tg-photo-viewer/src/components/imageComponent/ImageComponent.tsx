"use client";
import Image from 'next/image';
import { ImageData } from '../../customTypes/types.ts';
import styles from './imgComponent.module.css';

const ImageComponent: React.FunctionComponent<ImageData> = (props: ImageData): React.JSX.Element => {
	// required with "placeholder='blur'"
	const blurDataURL: string = `data:image/jpeg;base64,${props.blur_hash}`;

	const handleClick = (e: React.SyntheticEvent): void => {
		e.preventDefault();

		// TODO implementation (open a dialog modal)
	}
	return (
		<Image className={styles.img}
			src={props.urls.small}
			alt={props.alt_description}
			placeholder='blur'
			blurDataURL={blurDataURL}
			width={250}
			height={props.height * 250 / props.width}
		/>
	)

	/*
	return (
		<figure id={props.id} className={styles.figure}
			onClick={handleClick}
			onTouchEnd={handleClick}
		>
			<figcaption className={styles.figcaption}>
				{props.description}
			</figcaption>
			<Image
				className={styles.img}
				src={props.urls.small}
				alt={props.alt_description}
				fill={true}
				placeholder='blur'
				blurDataURL={blurDataURL}
			/>
		</figure>
	)
	 */
};

export default ImageComponent;
