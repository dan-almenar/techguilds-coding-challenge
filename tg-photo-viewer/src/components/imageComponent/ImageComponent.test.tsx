import { render, screen, fireEvent } from '@testing-library/react';
import { ImageData } from '../../customTypes/types.ts';
import ImageComponent from './ImageComponent.tsx';

describe('ImageComponent tests', () => {
	const mockData: ImageData = {
		id: 'test-image',
		width: 7753,
		height: 5155,
		blur_hash: "LJ8EuXWC8^M|ajozM_j=roaKS7kC",
		description: 'Test image description',
		alt_description: 'test image alt_description',
		urls: {
			full: "https://images.unsplash.com/photo-1682687982167-d7fb3ed8541d?crop=entropy&cs=srgb&fm=jpg&ixid=M3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw&ixlib=rb-4.0.3&q=85",
    		regular: "https://images.unsplash.com/photo-1682687982167-d7fb3ed8541d?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=M3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw&ixlib=rb-4.0.3&q=80&w=1080",
    		small: "https://images.unsplash.com/photo-1682687982167-d7fb3ed8541d?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=M3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw&ixlib=rb-4.0.3&q=80&w=400",
		},
		links: {
			html: "https://unsplash.com/photos/a-scuba-diver-swims-through-an-underwater-cave-yx7TJle8LhM",
    		download: "https://unsplash.com/photos/yx7TJle8LhM/download?ixid=M3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw",
		},
		likes: 677,
		user: {
			id: 'test user',
			username: 'test username',
			name: 'test name',
			portfolio_url: 'https://google.com'
		},
	};

	it ('container matches snapshot', () => {
		const { container } = render(<ImageComponent {...mockData} />, {});
		expect(container).toMatchSnapshot();
	})
	it('<figure> tag renders properly', () => {
		const { container } = render(<ImageComponent {...mockData} />, {});
		const figureTag = container.querySelector('figure');
		expect(figureTag).toMatchSnapshot();
	});

	it('<figcaption> renders properly', () => {
		const { container } = render(<ImageComponent {...mockData} />, {});
		const captionTag = container.querySelector('figcaption');

		expect(captionTag).toBeInTheDocument();
	});

	it('img tag renders property', () => {
		const { container } = render(<ImageComponent {...mockData} />, {});
		const img = container.querySelector('img');

		expect(img).toBeInTheDocument();
	});

	it('Image src property is properly set', () => {
		const { container } = render(<ImageComponent {...mockData} />, {});
		const img = container.querySelector('img');

		expect(img).toHaveAttribute('src');
	});

	it('Component properly responds to click event', () => {
		const { container } = render(<ImageComponent {...mockData} />, {});
		const figureTag = container.querySelector('figure');

		expect(() => {
			fireEvent.click(figureTag);
		}).not.toThrow();
	});

	it('Component handles touch events properly', () => {
		const { container } = render(<ImageComponent {...mockData} />, {});
		const figureTag = container.querySelector('figure');

		expect(() => {
			fireEvent.touchEnd(figureTag);
		}).not.toThrow();
	});
});

