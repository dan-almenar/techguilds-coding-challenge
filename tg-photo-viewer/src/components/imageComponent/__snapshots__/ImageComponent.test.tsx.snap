// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageComponent tests <figure> tag renders properly 1`] = `
<figure
  class="figure"
  id="test-image"
>
  <figcaption
    class="figcaption"
  >
    Test image description
  </figcaption>
  <img
    alt="test image alt_description"
    class="img"
    data-nimg="fill"
    decoding="async"
    loading="lazy"
    sizes="100vw"
    src="/_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=3840&q=75"
    srcset="/_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=3840&q=75 3840w"
    style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
  />
</figure>
`;

exports[`ImageComponent tests container matches snapshot 1`] = `
<div>
  <figure
    class="figure"
    id="test-image"
  >
    <figcaption
      class="figcaption"
    >
      Test image description
    </figcaption>
    <img
      alt="test image alt_description"
      class="img"
      data-nimg="fill"
      decoding="async"
      loading="lazy"
      sizes="100vw"
      src="/_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=3840&q=75"
      srcset="/_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1682687982167-d7fb3ed8541d%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DM3w1NDIzOTl8MXwxfGFsbHwxfHx8fHx8Mnx8MTcwNTQxOTAzOXw%26ixlib%3Drb-4.0.3%26q%3D80%26w%3D400&w=3840&q=75 3840w"
      style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
    />
  </figure>
</div>
`;
